### PR TITLE
Update clap and remove use of structopt in tools

### PR DESCRIPTION
**Pull request by stmcginnis**
**Issue number:**

Related #2373

**Description of changes:**

This updates our `tools/` binaries to use the latest version of the `clap` crate. It also drops the use of `structopt` as that project is now in maintenance mode:

> As clap v3 is now out, and the structopt features are integrated into (almost as-is), structopt is now in maintenance mode: no new feature will be added.

So this standardizes our out-of-distro tools to just use the latest `clap`.

Note: unfortunately we need to pull in two different versions of `clap` due to our use of `generate-readme`. That project is no longer maintained as well, so we will either need to find a replacement or stop using it to generate README files.

**Testing done:**

Built all binaries. Compared `$bin --help` output between the old and new versions to verify there were no non-aesthetic differences.

Ran `cargo make && cargo make ami` to make sure project could still be built and published.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: tools/pubsys/src/aws/ami/mod.rs [hunk 1]
```diff
@@ -30,49 +31,47 @@ use serde::{Deserialize, Serialize};
...
 
 const WARN_SEPARATOR: &str = "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!";
 
 /// Builds Bottlerocket AMIs using latest build artifacts
-#[derive(Debug, StructOpt)]
-#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+#[derive(Debug, Parser)]
 pub(crate) struct AmiArgs {
     /// Path to the image containing the os volume
-    #[structopt(short = "o", long, parse(from_os_str))]
```
Comment (jpculp) [inline tools/pubsys/src/aws/ami/mod.rs:43]: For some reason I thought we had to do something more for the OsStr, but maybe I was overthinking.

Comment (stmcginnis) [reply tools/pubsys/src/aws/ami/mod.rs:43]: I think that may have been the case in older versions with `structopt`, but it appears `clap` handles things correctly.

File: tools/infrasys/src/main.rs [hunk 2]
```diff
@@ -16,36 +17,33 @@ use std::collections::HashMap;
...
-    // Path to Infra.toml  (NOTE: must be specified before subcommand)
-    #[structopt(long, parse(from_os_str))]
+    // Path to Infra.toml (NOTE: must be specified before subcommand)
+    #[arg(long)]
     infra_config_path: PathBuf,
 
-    #[structopt(subcommand)]
-    subcommand: SubCommand,
+    #[command(subcommand)]
+    subcommand: SubCommands,
```
Comment (zmrow) [inline tools/infrasys/src/main.rs]: Nit: enums are usually named with the singular form of the word, as an enum can only be one of it's contained variants and is not a collection.

File: tools/infrasys/src/main.rs [hunk 3]
```diff
@@ -16,36 +17,33 @@ use std::collections::HashMap;
...
 }
 
-#[derive(Debug, StructOpt)]
-#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
```
Comment (zmrow) [inline tools/infrasys/src/main.rs:40]: I had a bit of trouble looking up `DeriveDisplayOrder`.  I'm assuming it doesn't exist now, but are we sure removing it doesn't change behavior?

File: tools/pubsys-setup/src/main.rs [hunk 4]
```diff
@@ -14,33 +15,33 @@ use std::fs;
...
 /// Helps you get started with credentials to make Bottlerocket images and repos.
-#[derive(Debug, StructOpt)]
+#[derive(Debug, Parser)]
 struct Args {
-    #[structopt(global = true, long, default_value = "INFO")]
+    #[arg(global = true, long, default_value = "INFO")]
     /// How much detail to log; from least to most: ERROR, WARN, INFO, DEBUG, TRACE
     log_level: LevelFilter,
 
-    #[structopt(long, parse(from_os_str))]
```
Comment (zmrow) [inline tools/pubsys-setup/src/main.rs:28]: Does clap handle this properly?  We previously had to use the `parse` argument to do this.

File: tools/pubsys/src/aws/publish_ami/mod.rs [hunk 5]
```diff
@@ -23,52 +24,49 @@ use std::collections::{HashMap, HashSet};
...
-    #[structopt(long, group = "mode")]
+    #[arg(long, group = "mode")]
     grant: bool,
     /// Revoke access from the given users/groups
-    #[structopt(long, group = "mode")]
+    #[arg(long, group = "mode")]
     revoke: bool,
 
-    #[structopt(flatten)]
+    #[command(flatten)]
```
Comment (zmrow) [inline tools/pubsys/src/aws/publish_ami/mod.rs:64]: Is `command` used here to get access to `flatten`?

File: tools/pubsys/src/aws/validate_ami/mod.rs [hunk 6]
```diff
@@ -10,34 +10,33 @@ use crate::aws::client::build_client_config;
...
+    #[arg(long)]
     expected_amis_path: PathBuf,
 
     /// Optional path where the validation results should be written
-    #[structopt(long, parse(from_os_str))]
+    #[arg(long)]
     write_results_path: Option<PathBuf>,
 
-    #[structopt(long, requires = "write-results-path")]
+    #[arg(long, requires = "write_results_path")]
```
Comment (zmrow) [inline tools/pubsys/src/aws/validate_ami/mod.rs:34]: Why the change from dashes to underscores?  Does this still work properly?

File: tools/infrasys/src/main.rs (hunk 3 above)
Comment (stmcginnis) [reply tools/infrasys/src/main.rs:40]: Yeah, docs are really light on this, especially when using the `derive` approach we do.

It turns out `clap` has made [`DeriveDisplayOrder` the default](https://github.com/clap-rs/clap/pull/3975), so we no longer need to set this for each arg struct.

File: tools/pubsys-setup/src/main.rs (hunk 4 above)
Comment (stmcginnis) [reply tools/pubsys-setup/src/main.rs:28]: Yes, it appears `clap` handles this correctly.

File: tools/pubsys/src/aws/publish_ami/mod.rs (hunk 5 above)
Comment (stmcginnis) [reply tools/pubsys/src/aws/publish_ami/mod.rs:64]: Yes, `clap` uses more specific attribute names to differentiate some of the options.

File: tools/pubsys/src/aws/validate_ami/mod.rs (hunk 6 above)
Comment (stmcginnis) [reply tools/pubsys/src/aws/validate_ami/mod.rs:34]: `clap` compares this value to the actual struct property, not the rendered argument name. So with the dashes it would error because there was no property found with that name. It needs to match the actual name on line 32.

File: tools/infrasys/src/main.rs (hunk 2 above)
Comment (stmcginnis) [reply tools/infrasys/src/main.rs]: Done!

File: tools/deny.toml [hunk 7]
```diff
@@ -86,6 +83,9 @@ skip-tree = [
     # TestSys uses a newer version of base64 and serde_yaml
     { name = "testsys-model", version = "=0.0.8" },
     { name = "bottlerocket-types", version = "=0.0.8" },
+
+    # generate-readme uses an old version of clap
+    { name = "clap", version = "2.34.0" }
```
Comment (jpculp) [inline tools/deny.toml]: Rather than skipping `clap`'s tree, I would just skip the tree for `generate-readme` itself.

Comment (stmcginnis) [reply tools/deny.toml]: True, good call!

**Review by ecpullen — APPROVED** [review]
LGTM

**Review by zmrow — APPROVED** [review]
:moon_cake:

**Review by jpculp — APPROVED** [review]

**Review by etungsten — APPROVED** [review]
